src/internal/storage/chunk/client.go

```diff
@@ -43,73 +50,105 @@ func NewClient(store kv.Store, mdstore MetadataStore, tr track.Tracker, name str
 
 // Create creates a new chunk from metadata and chunkData.
 // It returns the ID for the chunk
-func (c *Client) Create(ctx context.Context, md Metadata, chunkData []byte) (_ ID, retErr error) {
+func (c *trackedClient) Create(ctx context.Context, md Metadata, chunkData []byte) (_ ID, retErr error) {
+	if c.renewer == nil {
+		panic("client must be named to create chunks")
+	}
 	chunkID := Hash(chunkData)
 	var pointsTo []string
 	for _, cid := range md.PointsTo {
-		pointsTo = append(pointsTo, ObjectID(cid))
+		pointsTo = append(pointsTo, cid.TrackerID())
 	}
-	chunkOID := ObjectID(chunkID)
-	if err := dbutil.WithTx(ctx, c.mdstore.DB(), func(tx *sqlx.Tx) error {
-		if err := c.mdstore.SetTx(tx, chunkID, md); err != nil && err != ErrMetadataExists {
+	chunkOID := chunkID.TrackerID()
+	var needUpload bool
+	var gen uint64
+	if err := dbutil.WithTx(ctx, c.db, func(tx *sqlx.Tx) error {
+		if err := c.tracker.CreateTx(tx, chunkOID, pointsTo, c.ttl); err != nil {
+			return err
+		}
+		var ents []Entry
+		if err := tx.Select(&ents, `
+		SELECT chunk_id, gen
+		FROM storage.chunk_objects
+		WHERE uploaded = TRUE AND tombstone = FALSE AND chunk_id = $1`, chunkID); err != nil {
+			return err
+		}
+		if len(ents) > 0 {
+			needUpload = false
+			return nil
+		}
+		if err := tx.Get(&gen, `
+		INSERT INTO storage.chunk_objects (chunk_id, size)
+		VALUES ($1, $2)
+		RETURNING gen
+		`, chunkID, md.Size); err != nil {
 			return err
 		}
-		return c.tracker.CreateTx(tx, chunkOID, pointsTo, c.ttl)
+		needUpload = true
+		return nil
 	}); err != nil {
 		return nil, err
 	}
 	if err := c.renewer.Add(ctx, chunkOID); err != nil {
 		return nil, err
 	}
-	// TODO: need to check for existence of specific objects.
-	key := chunkKey(chunkID)
-	if exists, err := c.store.Exists(ctx, key); err != nil {
-		return nil, err
-	} else if exists {
+	if !needUpload {
 		return chunkID, nil
 	}
+	key := chunkKey(chunkID, gen)
 	if err := c.store.Put(ctx, key, chunkData); err != nil {
 		return nil, err
 	}
+	_, err := c.db.Exec(`
+	UPDATE storage.chunk_objects
+	SET uploaded = TRUE
+	WHERE chunk_id = $1 AND gen = $2
+	`, chunkID, gen)
+	if err != nil {
+		return nil, err
+	}
 	return chunkID, nil
 }
 
 // Get writes data for a chunk with ID chunkID to w.
-func (c *Client) Get(ctx context.Context, chunkID ID, cb kv.ValueCallback) (retErr error) {
-	key := chunkKey(chunkID)
+func (c *trackedClient) Get(ctx context.Context, chunkID ID, cb kv.ValueCallback) (retErr error) {
+	var gen uint64
+	err := c.db.Get(&gen, `
+	SELECT gen
+	FROM storage.chunk_objects
+	WHERE uploaded = TRUE AND tombstone = FALSE AND chunk_id = $1
+	LIMIT 1
+	`, chunkID)
+	if err != nil {
+		err = errors.Errorf("no objects for chunk %v", chunkID)
```

Comment:
It seems like this should be handling sql no rows and just returning the error otherwise.